src/Lights/Shadows/shadowGenerator.ts

```diff
@@ -656,10 +681,100 @@ export class ShadowGenerator implements IShadowGenerator {
         if (this._shadowMap2) {
             return this._shadowMap2;
         }
-
         return this._shadowMap;
     }
 
+    private _viewSpaceFrustums: Array<Array<Vector3>>;
+    private _viewSpaceBoundingSpheres: Array<BoundingSphere>;
+    private _CSMtransformMatrices: Array<Matrix>;
+    private _cascades = 1;
+    /**
+     * Gets the number of cascades used in CSM mode.
+     */
+    public get cascades(): number {
+        return this._cascades;
+    }
+    /**
+     * Sets the number of cacades used in CSM mode.
+	 * Limited to 2 to 4 cascade to reduce resource footprint
+     */
+    public set cascades(value: number) {
+        if (this._cascades === value || value < ShadowGenerator.MIN_CASCADES || value > ShadowGenerator.MAX_CASCADES) {
+            return;
+        }
+        this._cascades = value;
+        if (this._cascades >= ShadowGenerator.MIN_CASCADES && this._cascades <= ShadowGenerator.MAX_CASCADES) {
+            this.useCSM = true;
+        }
+    }
+
+    private _useCSM = false;
+    /**
+     * Gets the CSM mode.
+     */
+    public get useCSM(): boolean {
+        return this._useCSM;
+    }
+    /**
+     * Sets the CSM mode.
+     */
+    public set useCSM(value: boolean) {
+        if (this._useCSM === value) {
+            return;
+        }
+        this._useCSM = value;
+    }
+
+    private _csmFrustumLength: number;
+    /**
+     * Gets the csmFrustumLength value: furthest range of the frustum for the CSM mode.
+     */
+    public get csmFrustumLength(): number {
+        if (!this._scene || !this._scene.activeCamera) {
+            return 0;
+        }
+        return this._csmFrustumLength !== undefined ? this._csmFrustumLength : this._scene.activeCamera.maxZ;
+    }
+    /**
+     * Sets the csmFrustumLength: furthest range of the frustum for the CSM mode.
+     */
+    public set csmFrustumLength(value: number) {
+        if (!this._scene || !this._scene.activeCamera) {
+            return;
+        }
+        if (this._csmFrustumLength === value || value < this._scene.activeCamera.minZ || value > this._scene.activeCamera.maxZ) {
+            return;
+        }
+        this._csmFrustumLength = value;
+    }
+
+    private _cascadeShadowMaps: Nullable<Array<RenderTargetTexture>>;
+
+    /**
+     * Gets the RTT containing the cascaded shadow map (usually storing depth from the light point of view).
+	 * @param index index number of the cascaded shadow map
+     * @returns The render target texture if present otherwise, null
+     */
+    public getCSM(index: number): Nullable<RenderTargetTexture> {

```

Comment:
Not sure we need this one, we might simply all the time use the array to prevent extra validation in a loop for instance

Reply:
Thanks for the catch! A whole bunch of this were put in during testing. Now removed !